packages/auth/src/mfa/assertions/totp.ts

```diff
@@ -67,12 +84,162 @@ export class TotpMultiFactorGenerator {
+      typeof mfaSession.auth !== 'undefined',
+      AuthErrorCode.INTERNAL_ERROR
+    );
+    const response = await startEnrollTotpMfa(mfaSession.auth, {
+      idToken: mfaSession.credential,
+      totpEnrollmentInfo: {}
+    });
+    return TotpSecret.fromStartTotpMfaEnrollmentResponse(
+      response,
+      mfaSession.auth
+    );
   }
+
   /**
    * The identifier of the TOTP second factor: `totp`.
    */
   static FACTOR_ID = FactorId.TOTP;
 }
+
+export class TotpMultiFactorAssertionImpl
+  extends MultiFactorAssertionImpl
+  implements TotpMultiFactorAssertion
+{
+  constructor(
+    readonly otp: string,
+    readonly enrollmentId?: string,
+    readonly secret?: TotpSecret
+  ) {
+    super(FactorId.TOTP);
+  }
+
+  static _fromSecret(
+    secret: TotpSecret,
+    otp: string
+  ): TotpMultiFactorAssertionImpl {
+    return new TotpMultiFactorAssertionImpl(
+      (otp = otp),
+      undefined,
+      (secret = secret)
+    );
+  }
+
+  static _fromEnrollmentId(
+    enrollmentId: string,
+    otp: string
+  ): TotpMultiFactorAssertionImpl {
+    return new TotpMultiFactorAssertionImpl(
+      (otp = otp),
+      (enrollmentId = enrollmentId)
+    );
+  }
+
+  /** @internal */
+  _finalizeEnroll(
+    auth: AuthInternal,
+    idToken: string,
+    displayName?: string | null
+  ): Promise<FinalizeMfaResponse> {
+    _assert(
+      typeof this.secret !== 'undefined',
+      auth,
+      AuthErrorCode.ARGUMENT_ERROR
+    );
+    return finalizeEnrollTotpMfa(auth, {
+      idToken,
+      displayName,
+      totpVerificationInfo: this.secret.makeTotpVerificationInfo(this.otp)
+    });
+  }
+
+  /** @internal */
+  _finalizeSignIn(
+    _auth: AuthInternal,
+    _mfaPendingCredential: string
+  ): Promise<FinalizeMfaResponse> {
+    throw new Error('method not implemented');
+  }
+}
+
+/**
+ * Provider for generating a {@link TotpMultiFactorAssertion}.
+ *
+ * Stores the shared secret key and other parameters to generate time-based OTPs.
+ * Implements methods to retrieve the shared secret key, generate a QRCode URL.
+ * @public
+ */
+export class TotpSecret {
+  /**
+   * Constructor for TotpSecret.
+   * @param secretKey - Shared secret key/seed used for enrolling in TOTP MFA and generating otps.
+   * @param hashingAlgorithm - Hashing algorithm used.
+   * @param codeLength - Length of the one-time passwords to be generated.
+   * @param codeIntervalSeconds - The interval (in seconds) when the OTP codes should change.
+   */
+  private constructor(
+    readonly secretKey: string,
+    readonly hashingAlgorithm: string,
+    readonly codeLength: number,
+    readonly codeIntervalSeconds: number,
+    // TODO(prameshj) - make this public after API review.
+    // This can be used by callers to show a countdown of when to enter OTP code by.
+    private readonly finalizeEnrollmentBy: string,
+    private readonly sessionInfo: string,
+    private readonly auth: AuthInternal
+  ) {}
+
+  static fromStartTotpMfaEnrollmentResponse(
```

Comment:
For this public visibility method and others that aren't meant for use by developers, please prefix with `_`.

Reply:
Thanks! Done.